File: src/com/opengarden/test/server/dao/ChatMessageDao.java
package com.opengarden.test.server.dao;

import java.util.ArrayList;
import java.util.List;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.opengarden.test.server.model.ChatMessage;

/*
 * Don't have a database nor persisting messages on server,
 * but wanted to structure the code as if I did so it would
 * be easier to add
 */
public class ChatMessageDao implements ChatDao {
	private List<ChatMessage> mMessageList;
	private ChatMessage mChatMessage;

	public ChatMessageDao() {
		mMessageList = new ArrayList<>();
	}

	@Override
	public void storeChatMessage(ChatMessage message) {
		mChatMessage = message;
	}

	@Override
	public ChatMessage getChatMessage() {
		return mChatMessage;
	}

	@Override
	public void storeChatMessage(String chatJson) {
		Gson gson = new Gson();
		JsonElement jelem = gson.fromJson(chatJson, JsonElement.class);

		if (jelem.isJsonArray()) {
			JsonArray array = jelem.getAsJsonArray();
			if (array != null) {
				for (JsonElement je : array) {
					ChatMessage msg = new Gson().fromJson(je, ChatMessage.class);
					mMessageList.add(msg);
				}
			}
		} else {
			ChatMessage msg = gson.fromJson(jelem, ChatMessage.class);
			mMessageList.add(msg);
		}

		if (mMessageList.size() == 1) { // convenience
			mChatMessage = mMessageList.get(0);
		}
	}

	@Override
	public List<ChatMessage> getAllMessages() {
		return mMessageList;
	}

	@Override
	public void deleteAllMessages() {
		mMessageList.clear();
		mChatMessage = null;
	}
}

File: presentation/terminal/src/main/kotlin/ai/sterling/kchat/server/dao/ChatMessageDao.kt
package ai.sterling.kchat.server.dao

import ai.sterling.kchat.domain.chat.model.ChatMessage
import ai.sterling.kchat.domain.chat.persistence.ChatDao
import java.util.*

/*
 * Don't have a database nor persisting messages on server,
 * but wanted to structure the code as if I did so it would
 * be easier to add
 */
class ChatMessageDao : ChatDao {
    private var messageList: MutableList<ChatMessage>
    private var chatMessage: ChatMessage? = null

    init {
        messageList = ArrayList()
    }

    override fun insertChatMessage(message: ChatMessage): Boolean {
        chatMessage = message
        return true
    }

    override fun insertAllMessages(messageList: List<ChatMessage>): Boolean {
        this.messageList = messageList.toMutableList()
        if (messageList.size == 1) { // convenience
            chatMessage = messageList[0]
        }

        return true
    }

    override fun getAllMessages(): List<ChatMessage> {
        return messageList
    }

    override fun getChatMessage(id: Int): ChatMessage? {
        return messageList.filter {
            it.id == id
        }[0]
    }

    override fun deleteAllMessages(): Boolean {
        messageList.clear()
        chatMessage = null

        return true
    }
}